### PR TITLE
Interface to initialize environments and agents

### LINKED ISSUE #32: Registering all environments with gym
It will be easier if we can register all environments with gym and then use `gym.make`.

The problem with using `configs` function call is that the static analysis and debugging is obscure currently. The extra arguments for the environments can already be provided via the gym `register` method.

The general flow would look like this

```
env = gym.make('PommeFFA-v0')
env.set_agents(agent_class_list)
```

instead of the runtime evaluation that we do currently. The configurations are static and hence the need for those runtime function calls are now unnecessary when `gym` already provides a way to do that for us. Plus, this way of sending in strings for parsing is brittle because if the class interfaces change, everybody's code breaks. A better idea might be to provide explicit class list. (Of course, if you want an additional CLI like `run_battle.py`, the string parsing logic can happen there)

See #29 for a step in this direction. Comments welcome.

### LINKED ISSUE #37: Does not work in python 2.7-3.4 due to circular dependency
when `examples/docker-agent/run.py` is run, it eventually gets to importing `envs.utility` which imports `..characters` which imports `envs.utility` causing circular dependency. 

Circular dependencies are implicitly handled from 3.5+ so it works in 3.6. 
There are two ways of solving this. 

1. Refactor modules to avoid circular dependency
2. Hack fix

Here is the full trace when `examples/docker-agent/run.py` is executed.
```
Traceback (most recent call last):
  File "/home/ragha/playground/examples/docker-agent/run.py", line 3, in <module>
    import pommerman
  File "/home/ragha/playground/pommerman/__init__.py", line 5, in <module>
    from . import configs
  File "/home/ragha/playground/pommerman/configs.py", line 1, in <module>
    from . import envs
  File "/home/ragha/playground/pommerman/envs/__init__.py", line 1, in <module>
    from . import v0
  File "/home/ragha/playground/pommerman/envs/v0.py", line 15, in <module>
    from . import utility
  File "/home/ragha/playground/pommerman/envs/utility.py", line 8, in <module>
    from ..characters import Flame
  File "/home/ragha/playground/pommerman/characters.py", line 5, in <module>
    from .envs import utility
ImportError: cannot import name utility
```

This is the part where circular imports happen.
```
File "/home/ragha/playground/pommerman/envs/v0.py", line 15, in <module>
    from . import utility
  File "/home/ragha/playground/pommerman/envs/utility.py", line 8, in <module>
    from ..characters import Flame
  File "/home/ragha/playground/pommerman/characters.py", line 5, in <module>
    from .envs import utility
```

Ideally, we should refactor to avoid the circular dependency. Which option do we want to go with? Votes?

### LINKED ISSUE #41: Change characters.Agent --> characters.Bomber
This name is confusing because it's not meant to be controlled by players and is close to BaseAgent, which is meant to be controlled.

### PR DESCRIPTION
**Pull request by activatedgeek**
- [x] Compatible with OpenAI Gym (Fixes #32)
- [x] `pommerman.registry` for a list of all environments
- [x] Ability to create agents manually (and send to `pommerman.make`)
- [x] An internal helper routine to make agents from strings (not to be used externally because unstable API)
- [x] Add min Python version check (Temporarily Fixes #37)
- [x] Rename `character.Agent` to `character.Bomber` (Fixes #41 )
- [x] Handle `agent_id`s during `pommerman.make`

### Minimum Working Example
```python
import pommerman
from pommerman import agents

def main():
    # Print all possible environments in the Pommerman registry
    print(pommerman.registry)

    agent_list = [
        agents.SimpleAgent(),
        agents.SimpleAgent(),
        agents.RandomAgent(),
        agents.DockerAgent("pommerman/simple-agent", port=12345)
    ]
    env = pommerman.make('PommeFFA-v0', agent_list)

    for i_episode in range(1):
        state = env.reset()
        done = False
        while not done:
            env.render()
            actions = env.act(state)
            state, reward, done, info = env.step(actions)
        print('Episode {} finished'.format(i_episode))
    env.close()


if __name__ == '__main__':
    main()
```

### Considerations

All the agents now take the **class** (remember not the instance!) of the character (defaults to `characters.Bomber`).

### CODE REVIEW

**Comment by weseldridge** [comment]
Some good cleanups in there. I haven't tested it yet. 

I was going to mention #37 but looks like you're own it. 🙌 In conversations, he didn't seem to be too worried about it.

File: pommerman/configs.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ def ffa_v0():
         'max_steps': envs.utility.MAX_STEPS,
```
Comment (cinjon) [inline pommerman/configs.py]: Thanks for including this

File: pommerman/agents/player_agent.py [hunk 2]
```diff
@@ -1,34 +1,41 @@
...
         super(PlayerAgent, self).__init__(agent)
-        self._key_input = {'curr': 0}
 
-        if control == 'arrows':
-            self._key2act = {
+        ##
+        # @NOTE: DONOT move this import outside the constructor. It will
+        # not work in headless environments like a Docker container
+        # and prevents Pommerman from running
+        #
```
Comment (cinjon) [inline pommerman/agents/player_agent.py]: Thanks for catching this. Slight typo on the comment string.

File: pommerman/agents/base_agent.py [hunk 3]
```diff
@@ -1,8 +1,10 @@
+from .. import characters
+
 
 class BaseAgent:
```
Comment (cinjon) [inline pommerman/agents/base_agent.py]: With the change to characters.Bomber and the push to make it default everywhere, we should additionally change it from self._agent to self._character.

File: pommerman/__init__.py [hunk 4]
```diff
@@ -1,73 +1,40 @@
...
-                env_vars[agent_id][key] = value
+def make(config_id, agent_list, game_state_file=None):
+    assert config_id in registry
+    env = gym.make(config_id)
 
-    _agents = []
-    for agent_id, agent_string in enumerate(agent_string.split(',')):
-        agent_instance = make_agent(config_string, agent_string, agent_id, env_vars.get(agent_id))
-        _agents.append(agent_instance)
+    for idx, _agent in enumerate(agent_list):
```
Comment (cinjon) [inline pommerman/__init__.py]: idx --> id,
_agent --> agent.

File: pommerman/__init__.py [hunk 5]
```diff
@@ -1,73 +1,40 @@
...
+        assert isinstance(_agent, agents.BaseAgent)
+        # @NOTE: This is IMPORTANT so that the agent character is initialized
+        _agent.init_agent(idx, env.spec._kwargs['game_type'])
```
Comment (cinjon) [inline pommerman/__init__.py]: _kwargs should be renamed so that it's visible.

File: pommerman/__init__.py [hunk 6]
```diff
@@ -1,73 +1,40 @@
...
 
-    env.set_agents(_agents)
+    env.set_agents(agent_list)
     env.set_init_game_state(game_state_file)
     return env
 
 
 # Register environments with gym
-register()
+_register()
```
Comment (cinjon) [inline pommerman/__init__.py]: put this call after the function so that it's all together.

File: pommerman/__init__.py [hunk 7]
```diff
@@ -1,73 +1,40 @@
...
 import inspect
-import os
-import docker
 from . import configs
 from . import utility
 from . import agents
 
-client = docker.from_env()
-servers = os.environ.get('PLAYGROUND_BATTLE_SERVERS', ','.join(['http://localhost']*4)).split(',')
+# servers = os.environ.get('PLAYGROUND_BATTLE_SERVERS', ','.join(['http://localhost']*4)).split(',')
```
Comment (cinjon) [inline pommerman/__init__.py]: remove this; ill figure out what to do with it once the dust settles.

**Comment by cinjon** [comment]
lgtm.

File: pommerman/agents/player_agent.py [hunk 8]
```diff
@@ -1,34 +1,41 @@
...
-            self._key2act = {
+    def __init__(self, character=characters.Bomber, agent_control='arrows'):
+        super(PlayerAgent, self).__init__(character)
+
+        ##
+        # @NOTE: DO NOT move this import outside the constructor. It will
+        # not work in headless environments like a Docker container
+        # and prevents Pommerman from running.
+        #
+        from pyglet.window import key
```
Comment (lucasb-eyer) [inline pommerman/agents/player_agent.py]: Ah, makes sense, sorry for introducing this bug!

**Comment by lucasb-eyer** [comment]
You changed the signature of `make` but didn't adapt call-sites such as [`run_battle.py` here](https://github.com/MultiAgentLearning/playground/blob/master/pommerman/cli/run_battle.py#L34). I don't yet understand your changes enough to fix it myself.

**Pull request merged**